### PR TITLE
DEV-1055: Update binary on plugin start

### PR DESCRIPTION
**Pull request by boaz-codota**
Ref https://codota.atlassian.net/browse/DEV-1055

### CODE REVIEW

File: src/main/java/com/tabnine/binary/TabNineProcessFacade.java [hunk 1]
```diff
@@ -1,39 +1,40 @@
...
 
-    public static void setTesting(TabNineBinary tabNineBinary) {
-        TabNineProcessFacade.tabNineBinary = tabNineBinary;
+    public static void setTesting(BinaryFacade tabNineBinary) {
+        TabNineProcessFacade.binaryFacade = tabNineBinary;
     }
 
-    public static void create() throws IOException {
-        tabNineBinary.create();
+    public static void create() throws IOException, NoValidBinaryToRunException {
```
Comment (dimacodota) [inline src/main/java/com/tabnine/binary/TabNineProcessFacade.java:16]: where the exceptions are handled?

File: src/main/java/com/tabnine/binary/BinaryFacade.java [hunk 2]
```diff
@@ -9,24 +9,34 @@
 import java.nio.charset.StandardCharsets;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
 
-import static com.tabnine.StaticConfig.generateCommand;
+public class BinaryFacade {
+    private final BinaryRun tabNineFinder;
```
Comment (dimacodota) [inline src/main/java/com/tabnine/binary/BinaryFacade.java:15]: `finder`? maybe name it `runner`?

**Review by dimacodota — COMMENTED** [review]
LGTM, see my questions, good test! big like!

File: src/main/java/com/tabnine/binary/TabNineProcessFacade.java [hunk 3]
```diff
@@ -1,39 +1,40 @@
...
 
+import com.tabnine.DependencyContainer;
 import com.tabnine.exceptions.TabNineDeadException;
 
 import javax.annotation.Nonnull;
 import java.io.IOException;
 
 public class TabNineProcessFacade {
-    private static TabNineBinary tabNineBinary = new TabNineBinary();
+    private static BinaryFacade binaryFacade = DependencyContainer.instanceOfBinaryFacade();
```
[the remark this answers is not in the document]
Comment (boaz-codota) [reply src/main/java/com/tabnine/binary/TabNineProcessFacade.java:10]: Inject how?

File: src/main/java/com/tabnine/binary/TabNineProcessFacade.java (hunk 1 above)
Comment (boaz-codota) [reply src/main/java/com/tabnine/binary/TabNineProcessFacade.java:16]: Here: https://github.com/codota/tabnine-intellij/blob/5d9f121b8557e28ac6147c8b738fc2406d7ec24d/src/main/java/com/tabnine/binary/TabNineGateway.java#L59

File: src/main/java/com/tabnine/binary/TabNineProcessFacade.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dimacodota) [reply src/main/java/com/tabnine/binary/TabNineProcessFacade.java:10]: ctor? who is initializing the TabNineProcessFacade ?

**Review by dimacodota — APPROVED** [review]
LGTM

**Pull request merged**